### PR TITLE
Install and use `ruff format` instead of black for code formatting.

### PR DESCRIPTION
**Pull request by obi1kenobi**
Best to review one commit at a time, since two of the commits are 100% autogenerated changes from running `ruff format`:
- Install and use `ruff format` instead of black for code formatting.
- Output of `ruff format .` in the `langchain` package.
- Use `ruff format` in experimental package.
- Format changes in experimental package by `ruff format`.
- Manual formatting fixes to make `ruff .` pass.

### CODE REVIEW

**Review by baskaryan — COMMENTED** [review]
there's a root level Makefile as well that should be updated

**Comment by obi1kenobi** [comment]
Planning to remove the `black` cache setup and associated complexity from the lint script in a separate PR, just to limit the review complexity here.

**Review by baskaryan — APPROVED** [review]

**Comment by obi1kenobi** [comment]
Will resolve the conflict and merge.

File: libs/langchain/langchain/tools/e2b_data_analysis/unparse.py [hunk 1]
```diff
@@ -683,7 +683,7 @@ def _arguments(self, t):
                     first = False
                 else:
                     self.write(", ")
-                self.dispatch(a),
+                (self.dispatch(a),)
```
Comment (MichaReiser) [inline libs/langchain/langchain/tools/e2b_data_analysis/unparse.py:686]: It might be unintentional to use a tuple here.

Comment (obi1kenobi) [reply libs/langchain/langchain/tools/e2b_data_analysis/unparse.py:686]: Great catch, thank you!

Comment (obi1kenobi) [reply libs/langchain/langchain/tools/e2b_data_analysis/unparse.py:686]: Fixed in #12691, thanks again!

**Pull request merged**